serverside/zerotrace/example/tcp_handshake.go

```diff
@@ -0,0 +1,407 @@
+	if !exists {
+		return 0, errNoConnState
+	}
+	return connState.mss()
+}
+
+func (s *stateMachine) rttByTuple(t *fourTuple) (time.Duration, error) {
+	s.RLock()
+	defer s.RUnlock()
+
+	if t == nil {
+		return 0, errNilTuple
+	}
+
+	connState, exists := s.m[*t]
+	if !exists {
+		return 0, errNoConnState
+	}
+	return connState.rtt()
+}
+
+func (s *stateMachine) addAndRtt(p gopacket.Packet) (time.Duration, error) {
+	if err := s.add(p); err != nil {
+		return 0, err
+	}
+
+	tuple, err := pktToTuple(p)
+	if err != nil {
+		return 0, err
+	}
+	return s.rttByTuple(tuple)
+}
+
+// pktToTuple extracts the four-tuple from the given packet: source IP address,
+// source port, destination IP address, destination port.
+func pktToTuple(p gopacket.Packet) (*fourTuple, error) {
+	var srcAddr, dstAddr net.IP
+
+	// Are we dealing with IPv4 or IPv6?
+	if p.NetworkLayer().LayerType() == layers.LayerTypeIPv4 {
+		v4 := p.Layer(layers.LayerTypeIPv4).(*layers.IPv4)
+		srcAddr = v4.SrcIP
+		dstAddr = v4.DstIP
+		if v4.Protocol != layers.IPProtocolTCP {
+			return nil, errIPHasNoTCP
+		}
+	} else if p.NetworkLayer().LayerType() == layers.LayerTypeIPv6 {
+		v6 := p.Layer(layers.LayerTypeIPv6).(*layers.IPv6)
+		srcAddr = v6.SrcIP
+		dstAddr = v6.DstIP
+		if v6.NextHeader != layers.IPProtocolTCP {
+			return nil, errIPHasNoTCP
+		}
+	} else {
+		return nil, errors.New("not an IPv4 or IPv6 packet")
+	}
+
+	tcp := p.Layer(layers.LayerTypeTCP).(*layers.TCP)
+	return newFourTuple(
+		srcAddr, uint16(tcp.SrcPort),
+		dstAddr, uint16(tcp.DstPort),
+	), nil
+}
+
+// isSynSegment returns true if the given packet is a TCP segment that has
+// only its SYN flag set.
+func isSynSegment(p gopacket.Packet) bool {
+	var tcp *layers.TCP
+	if p.TransportLayer().LayerType() == layers.LayerTypeTCP {
+		tcp = p.Layer(layers.LayerTypeTCP).(*layers.TCP)
+	} else {
+		return false
+	}
+	// Only the SYN flag must be set.
+	if tcp.FIN || tcp.RST || tcp.PSH || tcp.URG || tcp.ECE || tcp.CWR || tcp.NS || tcp.ACK {
+		return false
+	}
+	return tcp.SYN
+}
+
+// isSynAckSegment returns true if the given packet is a TCP segment that has
+// only its SYN and ACK flags set.
+func isSynAckSegment(p gopacket.Packet) bool {
+	var tcp *layers.TCP
+	if p.TransportLayer().LayerType() == layers.LayerTypeTCP {
+		tcp = p.Layer(layers.LayerTypeTCP).(*layers.TCP)
+	} else {
+		return false
+	}
+	// Only the SYN/ACK flag must be set.
+	if tcp.FIN || tcp.RST || tcp.PSH || tcp.URG || tcp.ECE || tcp.CWR || tcp.NS {
+		return false
+	}
+	return tcp.SYN && tcp.ACK
+}
+
+// isAckSegment returns true if the given packet is a TCP segment that has only
+// its ACK flag set.
+func isAckSegment(p gopacket.Packet) bool {
+	var tcp *layers.TCP
+	if p.TransportLayer().LayerType() == layers.LayerTypeTCP {
+		tcp = p.Layer(layers.LayerTypeTCP).(*layers.TCP)
+	} else {
+		return false
+	}
+	// Only the ACK flag must be set.
+	if tcp.FIN || tcp.SYN || tcp.RST || tcp.PSH || tcp.URG || tcp.ECE || tcp.CWR || tcp.NS {
```

Comment:
Found a 🐞 with checking these flags. 

I found the following case where the ECN and CWR flags are being set by the VPN, presumably to indicate the capability for handling Explicit Congestion Notifications:
```
87	30.043184	<VPN_IP>	<SERVER_IP>	TCP	78	60707 → 443 [SYN, ECN, CWR] Seq=0 Win=65535 Len=0 MSS=1340 WS=64 TSval=1754955918 TSecr=0 SACK_PERM=1
88	30.043226	<SERVER_IP>	<VPN_IP>	TCP	74	443 → 60707 [SYN, ACK, ECN] Seq=0 Ack=1 Win=26847 Len=0 MSS=8961 SACK_PERM=1 TSval=2960511171 TSecr=1754955918 WS=128
89	30.403246	<VPN_IP>	<SERVER_IP>	TCP	66	60707 → 443 [ACK] Seq=1 Ack=1 Win=131456 Len=0 TSval=1754956384 TSecr=2960511171
```
And the webserver responds appropriately with the ECN flag set in the SYN-ACK packet. I think we should account for this, and accommodate it. 

I tried simply removing those flags from the checks on lines 289 and 305 but that doesn't do the trick. @NullHypothesis Would you be able to take a look into this? 🙏 

Btw, for the runs where this would have been an issue, we have the pcaps, plus since I logged the FourTuple, we know exactly which sets of packets SYN-ACK and ACK packets to extract this info from the pcap.

Reply:
Maybe we have to remove the check for `tcp.NS` (whose purpose I don't fully understand) as well?

Reply:
I was sure I was missing something else in the code in `tcp_handshake.go` because I tried: 
1) Minimal changes: removing CWR and ECE in isSynSegment and isSynAckSegment alone, and
2) Then, tried removing all non-essentials (i.e. other than the main 6 flags) including CWR, ECE, and NS from all three functions, neither seemed to do the trick 😕 
I'll try again, and do with some debug statements in a bit.